tools/seistech_utils/seistech_utils/src/api/shared.py

```diff
@@ -417,6 +420,132 @@ def create_uhs_download_zip(
     return zip_ffp
 
 
+def write_gms_download_data(
+    gms_result: sc.gms.GMSResult,
+    app: flask.app,
+    out_dir: str,
+    disagg_data: sc.disagg.EnsembleDisaggResult,
+    cs_param_bounds: sc.gms.CausalParamBounds = None,
+    prefix: str = None,
+):
+    prefix = "" if prefix is None else f"{prefix}_"
+
+    missing_waveforms = gms_result.gm_dataset.get_waveforms(
+        gms_result.selected_gms_ids, gms_result.site_info, out_dir
+    )
+    if len(missing_waveforms) > 0:
+        app.logger.info(
+            f"Failed to find waveforms for simulations: {missing_waveforms}"
+        )
+
+    if cs_param_bounds is not None:
+        # Mw and Rrup distribution plot
+        sc.plots.plt_gms_mw_rrup(
+            gms_result,
+            disagg_data.mean_values,
+            cs_param_bounds=cs_param_bounds,
+            save_file=Path(out_dir) / f"{prefix}gms_mw_rrup_plot.png",
+        )
+
+        # Pseudo acceleration response spectra plot
+        sc.plots.plt_gms_spectra(
+            gms_result,
+            save_file=Path(out_dir) / f"{prefix}gms_spectra_plot.png",
+        )
+
+        # IM distribution plots
+        sc.plots.plt_gms_im_distribution(gms_result, save_file=Path(out_dir))
+
+        # Disagg Distribution plots (Mw Distribution or Rrup distribution)
+        contribution_df_data = sr.get_default_causal_params(cs_param_bounds)[
+            "contribution_df"
+        ]
+        if contribution_df_data is not None:
+            sorted_rrup, rrup_cdf = utils.calc_cdf(
```

Comment:
Instead of doing the sorting here and the plotting function assuming that the passed in data is sorted, I suggest that the sorting is moved into the plotting function. Combined with the other comment above of taking a series instead of 2 separate lists the call from here then just becomes the following (for magnitude)
```
            sc.plots.plt_gms_disagg_distribution(
               cs_param_bounds.contr_df.loc[:, ["magnitude", "contribution"]].set_index("magnitude", drop=True),
                gms_result,
                "mag",
                cs_param_bounds=cs_param_bounds,
                save_file=Path(out_dir)
                / f"{prefix}gms_mag_disagg_distribution_plot.png",
            )
```
And you can remove line 460 & 464-470

Reply:
Did manage to sort within the plotting function. However, had to bring `calc_cdf` function to `plt_plotting.py` due to circular imports.